spec/visual/app.rb

```diff
@@ -42,7 +42,7 @@ def query_string
     @theme = theme_class.new(scope: '.codehilite')
     @comment_color = @theme.class.get_style(Rouge::Token::Tokens::Comment).fg
 
-    formatter_opts = { :line_numbers => params[:line_numbers] }
+    formatter_opts = { line_numbers: params[:line_numbers], line_table: true }
     formatter_opts[:inline_theme] = @theme if params[:inline]
 
     @formatter = Rouge::Formatters::HTMLLegacy.new(formatter_opts)
```

Comment:
Is this the reason you were editing the HTML legacy formatted? Why not just change the code here to switch based on whether it's a line table or not?

Reply:
Yes, this is the reason I changed the legacy formatter.
I found no compelling reason to *duplicate logic* used by the legacy formatter to use the proposed formatter here.

However, I'm open to better alternatives.

Reply:
I can't see a way to avoid some duplication of the internal logic of `HTMLLegacy` but since you're already writing code to transform `params` into `formatter_opts` so that `HTMLLegacy` can do its thing, you can eliminate that code and produce the result you want in fewer lines.

So, instead of this:

```ruby
    formatter_opts = { line_numbers: params[:line_numbers], line_table: true }
    formatter_opts[:inline_theme] = @theme if params[:inline]

     @formatter = Rouge::Formatters::HTMLLegacy.new(formatter_opts)
```

you'd write:

```ruby
    @formatter = param[:inline] ? Rouge::Formatters::HTMLInline.new(@theme) : Rouge::Formatters::HTML.new
    @formatter = param[:line_numbers] ? Rouge::Formatters::HTMLLineTable.new(@formatter) : @formatter
```

I can see why this might seem a bit perverse. Does it really matter? I don't think it's a massive deal but it feels wrong to me to be adding functionality to a class called `HTMLLegacy` to improve an internally facing web app. (I'll also note that getting rid of `HTMLLegacy` from the visual testing app seems like an added bonus.)

Reply:
This is basically what the constructor for the legacy formatter does..

My intention wasn't just to improve the internally facing app here, but also let end-users (of libraries that use Rouge) opt-in to the new formatter with a configuration parameter.

Hmm.. :thinking:

Reply:
Right, as I said:

1. I can't see a way to avoid duplication.

2. Adding functionality to a formatter with the word 'legacy' in its title seems contrary to the purpose of that formatter. Your formatter will be available to end-users, just not through `HTMLLegacy`.

@jneen Do you have a view?

Reply:
I agree with the first point. There is duplication. But in an app outside the library. So it doesn't matter.
Regarding the second, *end-users* won't be able to use this until middlemen like Jekyll or kramdown ships a wrapper. But I also agree that *legacy* shoudn't be upgraded..

I'll wait for @jneen opinion on this. Let's not have this block a Rouge 3.5.0 release..

Reply:
@jneen: Requesting your opinion on this discussion. Thanks

Reply:
Most other formatters that do things like this receive a formatter as an argument, would that be possible? The `span` method is public. We changed from using a super cursed options hash to building formatters out of other formatters, basically.

Reply:
...and for users that want more specific functionality, writing a formatter is not difficult. Gitlab, for example, uses a slight variant on the `HTMLLinewise` formatter - but since most of the logic is in the `span` method, there really isn't a whole lot of duplication.

Reply:
I would also prefer to update the visual spec app to use a non-legacy formatter than hack at the legacy formatter to make it work with the app.

Reply:
@jneen Thank you for your inputs. I have a few counter-arguments, though:

> Most other formatters that do things like this receive a formatter as an argument, would that be possible?

`HTMLLineTable` receives a `formatter` as well.

> for users that want more specific functionality, writing a formatter is not difficult

I agree. But what about the `end-users` of libraries like Jekyll and kramdown? They may not know Ruby at all. They usually depend on YAML configurations or fenced-code syntaxes to render highlighted code.

IMO, its improper to expect both Jekyll and kramdown  to ship their own flavors of formatters to cater *to their wide user-bases' varying interests*.

Reply:
> `HTMLLineTable` receives a `formatter` as well.

Then why does it need formatter options? And why does it have to instantiate a formatter? Are you sure you don't mean it receives a formatter class?

> But what about the `end-users` of libraries like Jekyll and kramdown?

I want to emphasize that output HTML formatting is extremely officially the responsibility of the developers of Jekyll and Kramdown, and not the highlighting library. We provide enough reasonable and extensible defaults for the developers of those libraries to make their own interfaces for their users.

It may seem extreme, but this approach comes from experience. We absolutely cannot deal with every user of every system that uses rouge asking for us to implement special formatting for their website.

Reply:
> IMO, its improper to expect both Jekyll and kramdown to ship their own flavors of formatters to cater to their wide user-bases' varying interests.

Basically, exactly this. So the developers of Jekyll and Kramdown (who *can* be expected to have familiarity with Ruby) can either write their own formatter (using pre-defined extension points like `#span` or `#safe_span` or `#token_lines`), or construct a formatter by combining other formatters. HTMLLineTable should be the latter, and should simply trust that the `#span` method of the passed-in formatter does the right thing.

For example

``` ruby
# a table with inline styles
Rouge::Formatters::HTMLLineTable.new(Rouge::Formatters::HTMLInline.new)

# a table with normal class names
Rouge::Formatters::HTMLLineTable.new(Rouge::Formatters::HTML.new)
```

Reply:
The developers of projects like Jekyll and kramdown are fully capable of using regular ruby programming to provide any kind of options they feel are necessary for their user base - it's not generally possible for us to express all possible ways of formatting HTML with a single options hash.

Reply:
Okay, having taken a more detailed look at the code, I can say plainly:

The HTMLLineTable implementation looks fine. +1 from me.
I am definitely against making any changes to HTMLLegacy.

Reply:
> it's not generally possible for us to express all possible ways of formatting HTML with a single options hash.

I'm beginning to think that we're on different pages here.
My intention with this PR is:
  - **Fix the markup rendered for code-blocks with linenos.**
  - Since I can't rehash existing `HTMLTable` formatter (already linked to the `HTMLLegacy` formatter) in `Rouge 3.x`, I set about proposing a new formatter that was meant to functionally work the same way as the existing `HTMLTable` but renders a different markup. Hence, lay-users could opt-in with a slight modification to their options which would get passed to the legacy formatter.

But looks like that won't be the case moving forward.
Thank you for all the inputs. I'll push the necessary changes a bit later..